### ISSUE TITLE
Support Alpine Linux

### ISSUE THREAD

**Issue by natemcmaster** [aspnet/Universe#833]
All of our native dependencies need to also be updated to support musl instead of glibc.

 - https://github.com/aspnet/libuv-package/issues/23
 - https://github.com/ericsink/SQLitePCL.raw/issues/192

We also need to generate an aspnetcore-runtime tarball for Alpine.

- [x] Upgrade SQLitePclRaw
- [ ] ~~Upgrade libuv~~ Deferred until 2.2. See https://github.com/aspnet/libuv-package/issues/23
- [ ]  Add functional tests for Alpine
- [x]  Update CI to produce an Alpine configuration
- [x]  Update dotnet/CLI to consume

**Comment by muratg** [comment]
SQLLitePCL.raw seems to be ready.  libuv-package task is assigned to @natemcmaster.

@glennc You had some concerns about this. Could you elaborate?  Tentatively putting this in the milestone so that we don't lose track of it.

**Comment by natemcmaster** [comment]
I believe @bricelam is currently assigned to pull in a new SQLitePCL.raw package for a separate reason. We can use https://github.com/aspnet/EntityFrameworkCore/issues/10724 to track that update.

**Comment by natemcmaster** [comment]
I've been poking at this for the last few days. The biggest challenge to implementing this at the moment is crossgen-ing the shared framework. This is difficult because

1. Installing dotnet via dotnet-install.sh into Alpine doesn't work yet. https://github.com/dotnet/cli/issues/8738 We can workaround for now by using curl/tar. 
1. MSBuild cannot compile a .csproj inside an Alpine container. https://github.com/Microsoft/msbuild/issues/3066
1. You can't run crossgen for Alpine on an Ubuntu machine because Ubuntu does not support the required libraries for musl-libc.

So, options:
1. Skip crossgen for now on Alpine Linux. (@DamianEdwards is this acceptable for a preview?)
2. Run crossgen without MSBuild. We could write MSBuild targets that generate shell scripts, but this must execute on an Ubuntu or Windows agent. Then, copy those scripts and files so crossgen can run inside a Docker container. (cc @JunTaoLuo )

**Comment by DamianEdwards** [comment]
How are they doing it for MS.NETCore.App?

**Comment by natemcmaster** [comment]
As far as I can tell, they are able to run the entire build inside an Alpine container. It's possible they are using a version of MSbuild that doesn't have this RAR bug: Microsoft/msbuild#3066 or that they don't compile any .csproj files in their build (that seems unlikely)

**Comment by natemcmaster** [comment]
I'm going to mark this as blocked for now on https://github.com/Microsoft/msbuild/issues/3066. It seems like https://github.com/Microsoft/msbuild/issues/3066 is critical enough (you can't even compile `dotnet new console` in Alpine) 🤞 it should get fixed.

**Comment by natemcmaster** [comment]
Unblocked. Found workarounds for now. It turns out MSBuild does not like .NET Core's invariant mode on Alpine :(. That can be worked around by installing libicu in the docker container.

**Comment by muratg** [comment]
@natemcmaster will that be in the .NET image as well, or just ours?

**Comment by natemcmaster** [comment]
For now, it's a workaround so we can _build_ the shared framework. It may not be necessary to _use_ the shared framework. 

Setting invariant mode by default in the microsoft/aspnetcore docker image is a separate question we should address with Rich, Michael, and Glenn. At the moment, I'm inclined to give invariant mode a shot so we can figure out what the problems will be. But it may not be a sensible default for customers.

**Comment by natemcmaster** [comment]
So, I looked at this myself. To disable invariant mode, a users needs to add this to their dockerfile.
```
RUN apk add --no-cache icu-libs
ENV DOTNET_SYSTEM_GLOBALIZATION_INVARIANT false
ENV LC_ALL en_US.UTF-8
ENV LANG en_US.UTF-8
```
`icu-libs` adds 30.4 MB to the image.

**Comment by natemcmaster** [comment]
Nightly builds for Alpine should be available on Docker Hub soon. https://github.com/aspnet/aspnet-docker/pull/390

Closing as remaining work listed in item description was punted to 2.2.0. Using aspnet/libuv-package#23 to track that.
